### PR TITLE
Windows fix linker call with quotes

### PR DESCRIPTION
**Pull request by Ingrater**
Fix calling the linker on windows when LINKCMD contains quotes.
Use same code path for 32-bit and 64-bit.

Examples of enviornment variables which cause issues:
LINKCMD=C:\\"Program Files (x86)"\\"Microsoft Visual Studio 14.0"\VC\bin\link.exe
LINKCMD="C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\link.exe"

Both are correctly quoted strings in terms of cmd but calling the linker will fail.

The rationale is this:
I have a LINKCMD with quotes and call dmd. Dmd reports that it can't call the linker and I should check PATH. I then run dmd with -v to get the command line dmd executes to call the linker. I copy this command line from the dmd output and paste it into a cmd. The command line executes just fine even though dmd just complained that it doesn't work.

### CODE REVIEW

**Comment by rainers** [comment]
LINKCMD is the file name of the linker, see https://dlang.org/dmd-windows.html. The quotes are not valid elements of the file name. Removing them is an artefact of C command line processing.
I seems odd to add text processing to this setting. The next guy will want to expand environment variables in it.
Can't you just cleanup your LINKCMD variable?

**Comment by Ingrater** [comment]
I could clean up my LINKCMD variable. But I find it strange. If you have a LINKCMD with quotes in it dmd will give you an error message saying that it can't run the linker. But when you copy the command that dmd claims to execute, it will work fine. Thats just unexpected behavior for a user.

File: src/link.d [hunk 1]
```diff
@@ -806,45 +807,49 @@ version (Windows)
...
-                STARTUPINFOA startInf;
-                startInf.dwFlags = STARTF_USESTDHANDLES;
-                startInf.hStdInput = GetStdHandle(STD_INPUT_HANDLE);
-                startInf.hStdOutput = GetStdHandle(STD_OUTPUT_HANDLE);
-                startInf.hStdError = GetStdHandle(STD_ERROR_HANDLE);
-                PROCESS_INFORMATION procInf;
-
-                BOOL b = CreateProcessA(null, cmdbuf.peekString(), null, null, 1, NORMAL_PRIORITY_CLASS, null, null, &startInf, &procInf);
-                if (b)
+                const(char)* start = cmd;
```
Comment (WalterBright) [inline src/link.d]: Some description of what the transformation is supposed to be would be nice, so it could be compared with the code.

File: src/link.d [hunk 2]
```diff
@@ -806,45 +807,49 @@ version (Windows)
         }
         // Normalize executable path separators, see Bugzilla 9330
         cmd = toWinPath(cmd);
-        version (CRuntime_Microsoft)
```
Comment (WalterBright) [inline src/link.d]: why is this removed?

Comment (Ingrater) [inline src/link.d]: So that there is one common code path for all cases and not two code paths depending on the cruntime used. The CreateProcess version will work with both c-runtimes the spawnlp version will not.

**Comment by Ingrater** [comment]
Before I start putting more work into this the central question should be answered first. Do we want this behavior like I propose or should we keep the old behavior and quotes in LINKCMD should be an error like proposed by rainers?

**Comment by WalterBright** [comment]
It's a good question. I don't know the answer, but I give a lot of weight to Rainer's opinion on it.

**Comment by rainers** [comment]
> I could clean up my LINKCMD variable. But I find it strange. If you have a LINKCMD with quotes in it dmd will give you an error message saying that it can't run the linker. But when you copy the command that dmd claims to execute, it will work fine. Thats just unexpected behavior for a user.

Yeah that's confusing. For COFF this is caused by omitting additional quotes when printing the command, though.

I just tried whether quotes worked with the spawn function for the optlink path: this fails, too.

Please note that the use of the environment variable LINKCMD is very limited as it is always set in the default sc.ini. So it only has effect when changing or disabling the configuration file with -conf=.

It is always a pain to work with spaces and quotes in paths as every program interpretes them differently. So I agree it would be good to be as permissive as possible. If we remove quotes in file names, we might want to add that to command line options, too. It should then be part of toWinPath.

One case that this PR might break: if someone used LINKCMD to pass additional options to the linker, removing quotes could make the options invalid.

BTW: just noticed that the default for the linker name was changed to "optlink" in the MSCOFF case, not the OMF case! The executable is still distributed as "link.exe", though.
